### ISSUE TITLE
dynamic dataset update

### ISSUE THREAD

**Issue by lucapaganotti** [Unidata/tds#466]
To report a non-security related issue, please provide:

* the version of the software with which you are encountering an issue
** tds 5.4
* environmental information (i.e. Operating System, compiler info, java version, python version, etc.)
** running in a docker container
** docker version 24.0.7
* a description of the issue with the steps needed to reproduce it

Good morning,

we are trying to setup a dynamic dataset that gets updated every 5 minutes. We wrote an nmcl file describing the aggregation, we are able to use joinExisting but it seems that in this case the recheckEvery attribute does not work (or we are missing something)
Trying to use joinNew we get errors.

This is the ncml file that describes the aggregation:

```
<?xml version="1.0" encoding="UTF-8"?>
<netcdf xmlns="http://www.unidata.ucar.edu/namespaces/netcdf/ncml-2.2">
  <aggregation dimName="time" type="joinExisting" recheckEvery="30 min">
    <scan location="/usr/local/tomcat/content/thredds/public/data/arpa/realtime" dateFormatMark="DES_VOL_CAP_#yyyyMMdd_HHmm" suffix="DES_VOL_CAP_20??????_????.nc" subdirs="true" />
  </aggregation>
  <remove name="start_time" type="variable" />
  <remove name="stop_time" type="variable" />
  <remove name="time_bounds" type="variable" />
</netcdf>
```
with this file we are able to view data but the dynamic dataset doesn't get updated.
When we change the aggregation type to joinNew we get errors in tds stating that:

Variable time already exists, but is not a scalar (suitable for aggregating as a coordinate)

We are using netcdf files that all have the same header structure, for example for one of them I get these informations for the time variable via ncdump -h:

```
netcdf DES_VOL_CAP_20240206_0000 {
dimensions:
    time = 1 ;
    bounds = 2 ;
    x0 = 680 ;
    y0 = 680 ;
    z0 = 10 ;
variables:
    double time(time) ;
        time:standard_name = "time" ;
        time:long_name = "Data time" ;
        time:units = "seconds since 1970-01-01T00:00:00Z" ;
        time:Axis = "T" ;
        time:bounds = "time_bounds" ;
        time:comment = "2024-02-06T00:03:20Z" ;

....
```

Is the problem due to the double data type assigned to the time dimension that insted should be for example int64? Or are we missing some other configuration detail?

Another question I have is where to find informations about correctly indexing geotiff files, we made some test on it but without getting any meaningful result.

Thank you for any answer.

**Comment by haileyajohnson** [comment]
Hello! `joinExisting` is correct, and time as `doubles` should not be problematic. What service(s) are you using to access this aggregation?

**Comment by WeatherGod** [comment]
Also, specifically, are you failing to see new files, or are you failing to see updates to existing files?

**Comment by tdrwenski** [comment]
Thanks for the extra info @lucapaganotti! WMS (which is used by Godiva3) had several issues in version 5.4 that we have recently fixed. This includes an issue that datasets would never get updated once they were cached, which sounds like what you are encountering. Can you test if it works as expected in today's thredds-docker 5.5-SNAPSHOT image?

**Comment by lucapaganotti** [comment]
Ok, nice to know, I'll try the 5.5 snapshot ASAP.

**Comment by tdrwenski** [comment]
Hi @lucapaganotti, we have pushed a new 5.5-SNAPSHOT image to the unidata dockerhub, so you should be able to just do ```
docker image pull unidata/thredds-docker:5.5-SNAPSHOT
```

**Comment by lucapaganotti** [comment]
Hello Tara,
I tried the 5.5-SNAPSHOT yesterday and unfortunately I have the same behaviour. 
I've stopped the 5.4 container, pulled the 5.5-SNAPSHOT image and changed the docker-compose.yml file in order to use the new image.
No other changes were made to the yaml file so we have the same volumes mounted as before.
What happens is that dataset metadata seem to be updated only at the very first time we open a .ncml file in thredds and try to view the maps using godiva.
We have 8 .ncml files, let's say that I opened the very first of them, then I waited for some time after some new .nc files were added to the right folder to view if the dataset time range displayed by godiva got updated but that did not happen.
I then checked a second .ncml file and also for this the same thing happened, it seems that the cache is updated only at the very first time I access the dataset represented by the .ncml aggregation, from that moment on no update happens anymore. 
I checked the object cache in thredds admim/debug interface and I noticed that the .ncml files I opened remain locked until I stop and run again the thredds container.
Are there some other checks I can do in order to give you more information about this issue?
 
I'm not an experienced/skilled java developer but I was wondering if it could be useful for me to get thredds code and work on it. 
I'm a c++ developer and I use eclipse for some of my projects, is there an already setup eclipse project to try to build/debug thredds in some way or do you have any hint to setup a development environment? I will check thredds docs to understand the building process.

Thanks for your work and for any answer.

**Comment by tdrwenski** [comment]
Also, just to be sure we are both testing the same snapshot-- can you check that the "Built-on" date you get when doing:
```
docker run --rm -a stdout --entrypoint cat unidata/thredds-docker:5.5-SNAPSHOT /usr/local/tomcat/webapps/thredds/META-INF/MANIFEST.MF
```
is at least `2024-03-02T04:38:12+0000`? Thanks!

**Comment by lucapaganotti** [comment]
Hi Tara,

yes, I do not see my aggregations updating also after the recheckEvery
setup value.
It seems that OpeNDAP correctly updates datetime aggregations' intervals, I
do not know about NCSS, I need to recheck the two services.

Thank you for your answer.
Have a nice day.

----------------------------------------------------------------
-- Dott. Ing. Luca Paganotti
-- Via Baroffio 4
-- 21040 Vedano Olona (VA)
-- 393 1346898
----------------------------------------------------------------
-- softech s.r.l. email:
-- ***@***.***
-- ***@***.***
-- https://github.com/lucapaganotti
-- sourceforge email:
-- ***@***.***
-- skype name: luca.paganotti
[image: http://it.linkedin.com/in/lucapaganotti]
<http://it.linkedin.com/in/lucapaganotti>
-- ---------------------------------------------------------------
-- Mistakes are portals of discovery - JAAJ
--- --------------------------------------------------------------

**Comment by lucapaganotti** [comment]
And yes I issued the docker run command you suggested, as you can see in the attached screenshot I get the Built-on value as you expect it to be.

![Schermata del 2024-03-15 10-59-25](https://github.com/Unidata/tds/assets/6130461/7ddae2bd-21e2-4cf2-8d62-7fdfa8281e1f)

**Comment by WeatherGod** [comment]
A couple of thoughts:

1. Different services sometimes maintain their own cache. .So, when you are checking if a service is showing updated data, did you check twice for each service (once before and once after the recheckEvery amount of time)?
2. Another thing I wonder is if the problem is with godiva rather than WMS. It would be useful to check what the WMS Capabilities document for the dataset says before and after the recheckEvery period. See my investigation into a very similar problem in #448.

**Comment by tdrwenski** [comment]
Hi @lucapaganotti, thanks for all the extra info! It sounds like NCSS and OPeNDAP are working as expected (getting the newer times). As @WeatherGod commented, it could be nice to test with WMS GetCapabilities just to rule out weird stuff happening with Godiva itself.

I did notice in your first screenshot with the docker built-on date that thredds-docker:5.4 is running. Just to be completely sure, were the rest of the tests performed with the 5.5-SNAPSHOT, not accidentally with the 5.4 version?

Would you be able to share your docker-compose.yml? I am just wondering, for instance, whether you are mounting your cache directories or they are inside the container. In addition, have you changed any settings in threddsConfig.xml? If so can you send that as well?

Thanks!

**Comment by WeatherGod** [comment]
To be clear, there are no WMS cache setup parameters. They should all use the same configurations. I just meant that as an implementation detail, some services maintain some sort of cache that might not be shared with other services. Therefore, when debugging this problem, it is better to treat the services as if they have their own independent caches in memory.

**Comment by tdrwenski** [comment]
Thanks for the extra info. Is it working as expected with 5.5-SNAPSHOT? In the GetCapabilities you can look at for instance
```
        <Dimension name="time" units="ISO8601" multipleValues="true" current="true" default="2024-03-19T07:45:00.000Z">
            2024-03-16T00:00:00.000Z/2024-03-19T07:45:00.000Z/PT5M
        </Dimension>
```
to see if the time range is getting updated when new files are added and the recheck time has passed.

**Comment by lucapaganotti** [comment]
Hi @WeatherGod and @tdrwenski,
we have checked all the relevant behaviour for our test cases and now I'm happy to say that all our dynamic datasets are updating correctly about every 10 minutes using the 5.5-SNAPSHOT and the finally corrected setup (as we are seeing for OpenDAP, NCSS Godiva3 viewer etc...), I'm very thankful for your kind help and sorry to have caused some noise here ... It's very likely I will need your help again in the future for some customizations of tds interface. I wish to be able to ask again.

@tdrwenski , yes we checked also the GetCapabilities time dimension to view if it was correctly updated and so it was.

Thank you all, have a nice day.
